tests/samplers_tests/test_samplers.py

```diff
@@ -1047,11 +1048,17 @@ def test_cache_is_invalidated(
     n_jobs: int, relative_sampler_class: Callable[[], BaseSampler]
 ) -> None:
     sampler = relative_sampler_class()
+    original_before_trial = sampler.before_trial
+
+    def mock_before_trial(study: Study, trial: FrozenTrial) -> None:
+        assert study._thread_local.cached_all_trials is None
+        original_before_trial(study, trial)
+
+    sampler.before_trial = Mock(side_effect=mock_before_trial)
```

Comment:
This line causes the "tests/samplers_tests/test_samplers.py:1057: error: Cannot assign to a method  [method-assign]" mypy error.

I think [`mock.patch.object`](https://docs.python.org/3.13/library/unittest.mock.html) is a better way to replace a class method.

Example:
```python
import unittest.mock


class C:
    def method(self):
        print("hello")

def mock_method():
    print("hi")

c = C()

c.method()
with unittest.mock.patch.object(C, 'method', unittest.mock.Mock(side_effect=mock_method)):
    c.method()
```
Output:
```
hello
hi
```